extensions/ql-vscode/src/debugger/debug-session.ts

```diff
@@ -0,0 +1,587 @@
+const QUERY_THREAD_ID = 1;
+
+/** The user-visible name of the query evaluation thread. */
+const QUERY_THREAD_NAME = "Evaluation thread";
+
+/**
+ * An in-process implementation of the debug adapter for CodeQL queries.
+ *
+ * For now, this is pretty much just a wrapper around the query server.
+ */
+export class QLDebugSession extends LoggingDebugSession implements Disposable {
+  private state: State = "uninitialized";
+  private terminateOnComplete = false;
+  private args: CodeQLProtocol.LaunchRequest["arguments"] | undefined =
+    undefined;
+  private tokenSource: CancellationTokenSource | undefined = undefined;
+  private queryRun: CoreQueryRun | undefined = undefined;
+  private lastResult:
+    | CodeQLProtocol.EvaluationCompletedEvent["body"]
+    | undefined = undefined;
+
+  constructor(
+    private readonly queryStorageDir: string,
+    private readonly queryRunner: QueryRunner,
+  ) {
+    super();
+  }
+
+  public dispose(): void {
+    this.cancelEvaluation();
+  }
+
+  protected dispatchRequest(request: Protocol.Request): void {
+    // We just defer to the base class implementation, but having this override makes it easy to set
+    // a breakpoint that will be hit for any message received by the debug adapter.
+    void queryServerLogger.log(`DAP request: ${request.command}`);
+    super.dispatchRequest(request);
+  }
+
+  private unexpectedState(response: Protocol.Response): void {
+    this.sendErrorResponse(
+      response,
+      ERROR_UNEXPECTED_STATE,
+      "CodeQL debug adapter received request '{_request}' while in unexpected state '{_actualState}'.",
+      {
+        _request: response.command,
+        _actualState: this.state,
+      },
+    );
+  }
+
+  protected initializeRequest(
+    response: Protocol.InitializeResponse,
+    _args: Protocol.InitializeRequestArguments,
+  ): void {
+    switch (this.state) {
+      case "uninitialized":
+        response.body = response.body ?? {};
+        response.body.supportsStepBack = false;
+        response.body.supportsStepInTargetsRequest = false;
+        response.body.supportsRestartFrame = false;
+        response.body.supportsGotoTargetsRequest = false;
+        response.body.supportsCancelRequest = true;
+        response.body.supportsTerminateRequest = true;
+        response.body.supportsModulesRequest = false;
+        response.body.supportsConfigurationDoneRequest = true;
+        response.body.supportsRestartRequest = false;
+        this.state = "initialized";
+        this.sendResponse(response);
+
+        this.sendEvent(new InitializedEvent());
+        break;
+
+      default:
+        this.unexpectedState(response);
+        break;
+    }
+  }
+
+  protected configurationDoneRequest(
+    response: Protocol.ConfigurationDoneResponse,
+    args: Protocol.ConfigurationDoneArguments,
+    request?: Protocol.Request,
+  ): void {
+    super.configurationDoneRequest(response, args, request);
+  }
```

Comment:
Do we need this method implementation? It seems to match `super.configurationDoneRequest` exactly.

Reply:
Nope. I must have added it to set a breakpoint more easily. Removed.